### PR TITLE
feat(cfloat): add quire and fused dot product support for cfloat

### LINKED ISSUE #547: feat(cfloat): add quire and fused dot product support for cfloat
## Parent Issue
Part of #345 — Explore generalization of the Fused Dot Product across static number systems

### Goal
Add quire accumulation and FDP (fused dot product) support for cfloat. cfloat already uses blocktriple as its arithmetic engine, making it the easiest new type to support.

### Design
- `quire_mul(cfloat, cfloat)` bridge: calls `normalize()` on each operand (produces blocktriple), multiplies, returns unrounded blocktriple product
- `fdp()` and `fdp_stride()` free functions guarded by `enable_if_cfloat`
- Uses the generalized `quire<cfloat<...>>` from step 2

### Acceptance Criteria
- [ ] `quire_mul()` for cfloat produces unrounded products
- [ ] `fdp()` round-trips correctly vs long double reference
- [ ] Cancellation-heavy test vectors pass
- [ ] Builds and tests with gcc and clang

### Files
- `include/sw/universal/number/cfloat/fdp.hpp` (new)
- `static/quire/cfloat/` test directory (new)

Relates to #345

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary
- Implements `quire_mul()`, `fdp()`, `fdp_stride()`, and `fdp_qc()` for cfloat via the generalized quire
- Fixes quire `accu_base_offset` to use blocktriple's compile-time radix instead of MSB position, correctly handling MUL overflow products
- Updates generalized_quire tests to use `quire_mul` products instead of raw `blocktriple(float)` constructors

## Changes
- `include/sw/universal/number/cfloat/fdp.hpp` — new: quire_mul and FDP free functions
- `include/sw/universal/number/quire/quire_impl.hpp` — fix: radix-based bit placement in accu_base_offset
- `static/quire/api/cfloat_fdp.cpp` — new: comprehensive FDP test suite
- `static/quire/api/generalized_quire.cpp` — updated: use quire_mul for blocktriple tests

## Test Results
| Target | gcc build | gcc test | clang build | clang test |
|--------|-----------|----------|-------------|------------|
| quire_generalized_quire | OK | PASS | OK | PASS |
| quire_cfloat_fdp | OK | PASS | OK | PASS |

## Test plan
- [x] Fast CI passes (gcc + clang CI_LITE)
- [x] Promote to ready when satisfied: `gh pr ready NNN`

Resolves #547

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added fused dot product (FDP) for cfloat with quire-based accumulation: unit-stride, strided, and quire-continuation variants for more accurate vector dot products.

* **Refactor**
  * Stabilized quire accumulation offset logic using a fixed radix-based placement to improve robustness and avoid overflow misplacement.

* **Tests**
  * Added extensive FDP test coverage: correctness, special values, limb/carry boundaries, strided/continuation stress, accuracy and cancellation cases across configurations.

* **Chores**
  * Expanded demos and examples for block-based numeric operations.

### CODE REVIEW

File: include/sw/universal/number/cfloat/fdp.hpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	if (lhs.iszero() || rhs.iszero()) return product;  // product is zero
+	if (lhs.isinf() || lhs.isnan() || rhs.isinf() || rhs.isnan()) {
+		product.setnan();
+		return product;
+	}
+
+	lhs.normalizeMultiplication(a);
+	rhs.normalizeMultiplication(b);
+	product.mul(a, b);
+	return product;
```
Comment (Ravenwater) [inline include/sw/universal/number/cfloat/fdp.hpp:49]: the product of two cfloats with fbits = (nbits - 1 - es) mantissa bits is 2*(fbits + 1). By declaring the result type of quire_mul blocktriple<fbits, MUL, bt>, do we get all the mantissa bits? quire_mul should return an unrounded multiply. Can you double check this is not rounding away any bits.

File: include/sw/universal/number/cfloat/fdp.hpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+/// Fused dot product with quire continuation.
+/// Accumulates products into an existing quire without resolving.
+template<typename Qy, typename Vector>
+void fdp_qc(Qy& sum_of_products, size_t n,
+            const Vector& x, size_t incx,
+            const Vector& y, size_t incy,
+            std::enable_if_t<is_cfloat<typename Vector::value_type>, int> = 0) {
+	size_t ix, iy;
+	for (ix = 0, iy = 0; ix < n && iy < n; ix += incx, iy += incy) {
+		sum_of_products += quire_mul(x[ix], y[iy]);
```
Comment (Ravenwater) [inline include/sw/universal/number/cfloat/fdp.hpp:154]: How does sum_of_products receive an unrounded blocktriple<2*(fbits+1), MUL>, is there a custom overload?

File: static/quire/api/cfloat_fdp.cpp [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+int TestQuireMul() {
+	int nrOfFailedTestCases = 0;
+
+	using Scalar = cfloat<32, 8, uint32_t, true, false, false>;
+
+	{
+		Scalar a(3.0f), b(5.0f);
+		auto product = quire_mul(a, b);
+		// product should represent 15.0 exactly
+		if (product.iszero() || product.isnan() || product.isinf()) {
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp]: this is a silly test, we input 3.0 and 5.0, so none of these results are expected. Testing for them ruins the flow of the test. Should be removed.

File: static/quire/api/cfloat_fdp.cpp [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+			std::cerr << "FAIL: quire_mul(3, 5) should be normal\n";
+			++nrOfFailedTestCases;
+		}
+		// accumulate in quire and extract
+		quire<Scalar> q;
+		q += product;
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp:84]: this is not a quire mul test, this is a quire load test. This is a weak test. quire loading and saving should focus on  more interesting cases relative to limb size, radix, and the value below and above the radix. Rewrite this test to be more useful.

File: static/quire/api/cfloat_fdp.cpp [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		double result = q.convert_to<double>();
+		if (result != 15.0) {
+			std::cerr << "FAIL: quire_mul(3, 5) should give 15.0, got " << result << '\n';
+			++nrOfFailedTestCases;
+		}
+	}
+
+	// Test with negative operands
+	{
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp:95]: definitely expand the testing of loading of different negative values

File: static/quire/api/cfloat_fdp.cpp [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+		if (result != -14.0) {
+			std::cerr << "FAIL: quire_mul(-2, 7) should give -14.0, got " << result << '\n';
+			++nrOfFailedTestCases;
+		}
+	}
+
+	// Test with zero
+	{
+		Scalar a(0.0f), b(42.0f);
+		auto product = quire_mul(a, b);
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp:202]: in addition to loading a quire_mul, we should also test loading a regular cfloat, and different blocktriple types, REP, ADD, MUL, DIV.

File: static/quire/api/cfloat_fdp.cpp [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+		if (!product.iszero()) {
+			std::cerr << "FAIL: quire_mul(0, 42) should be zero\n";
+			++nrOfFailedTestCases;
+		}
+	}
+
+	// Test with NaN
+	{
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp:140]: need to expand with loading special values: +-inf, (sq)NaN, +-0, through cfloat and blocktriple

File: static/quire/api/cfloat_fdp.cpp [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+// Test basic fdp with simple vectors
+int TestBasicFdp() {
+	int nrOfFailedTestCases = 0;
+
+	using Scalar = cfloat<32, 8, uint32_t, true, false, false>;
+
+	// dot product: [1, 2, 3, 4] . [5, 6, 7, 8] = 5 + 12 + 21 + 32 = 70
+	std::vector<Scalar> a = { Scalar(1.0f), Scalar(2.0f), Scalar(3.0f), Scalar(4.0f) };
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp]: too weak, focus on better test cases that create boundary conditions for the limb-based carry and borrow propagation. Also, FDP is to avoid catastrophic cancellation, so you need to have test cases where this is captures. Say a 1024 element vectors where we have 1023 small values that yield a value that is equivalent to the LSB of the 1024th element, so that the 1023 small values contribute the LSB to the final result.

File: static/quire/api/cfloat_fdp.cpp [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+	}
+
+	return nrOfFailedTestCases;
+}
+
+// Test fdp_stride
+int TestFdpStride() {
+	int nrOfFailedTestCases = 0;
+
+	using Scalar = cfloat<32, 8, uint32_t, true, false, false>;
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp:843]: need to have stronger test cases. shoot for at least 10 different carry and borrow cases.

File: static/quire/api/cfloat_fdp.cpp [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+
+// Test fdp_qc (quire continuation)
+int TestFdpQc() {
+	int nrOfFailedTestCases = 0;
+
+	using Scalar = cfloat<32, 8, uint32_t, true, false, false>;
+
+	std::vector<Scalar> x = { Scalar(1.0f), Scalar(2.0f), Scalar(3.0f) };
+	std::vector<Scalar> y = { Scalar(4.0f), Scalar(5.0f), Scalar(6.0f) };
+	// expected: 1*4 + 2*5 + 3*6 = 4 + 10 + 18 = 32
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp]: need to have much stronger test cases. The quire continuation should trigger very strong carry and borrow propagation cases, so that we can inject these cases clearly. Shoot for at least 10 such stress configurations.

File: static/quire/api/cfloat_fdp.cpp [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+		std::cerr << "FAIL: fdp_qc continuation should give 50.0, got " << result << '\n';
+		++nrOfFailedTestCases;
+	}
+
+	return nrOfFailedTestCases;
+}
+
+// Test FDP with cancellation-heavy vectors
+// The key advantage of FDP: it avoids catastrophic cancellation
+int TestCancellationHeavy() {
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp]: exactly, but this needs to be much stronger. Shoot for at least 10 stress test cases.

File: static/quire/api/cfloat_fdp.cpp [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+	// Build vectors where naive fp32 accumulation would lose precision
+	constexpr size_t N = 100;
+	std::vector<Scalar> x(N), y(N);
+	long double reference = 0.0L;
+	for (size_t i = 0; i < N; ++i) {
+		float xv = static_cast<float>(i + 1) * 0.1f;
+		float yv = static_cast<float>(N - i) * 0.1f;
+		x[i] = Scalar(xv);
+		y[i] = Scalar(yv);
+		// compute reference using long double from the cfloat-rounded values
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp]: do not use long double in test cases, as many of the cross platform targets do not have long double and it creates a lot of variability. depend on double as that is reliable.

File: static/quire/api/cfloat_fdp.cpp [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+		Scalar result = fdp(x, y);
+		double d = double(result);
+		if (d != 23.0) {
+			std::cerr << "FAIL: fp16 fdp should be 23.0, got " << d << '\n';
+			++nrOfFailedTestCases;
+		}
+	}
+
+	// fp8-equivalent (quarter precision)
+	// cfloat<8,2> has fbits=5, which gives quire_traits mbits=6, escale=22
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp]: needs to be stronger. take the core three fp8 configurations es=5, es=4, es=3.

File: static/quire/api/cfloat_fdp.cpp [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+		std::vector<Scalar> y = { Scalar(1.0f), Scalar(1.0f) };
+		// 1*1 + 2*1 = 3
+		Scalar result = fdp(x, y);
+		double d = double(result);
+		if (d != 3.0) {
+			std::cerr << "FAIL: fp8 fdp should be 3.0, got " << d << '\n';
+			++nrOfFailedTestCases;
+		}
+	}
+
```
Comment (Ravenwater) [inline static/quire/api/cfloat_fdp.cpp:1373]: Add odd configurations, cfloat<20,8>, cfloat<16,8>, cfloat<12, 5>, cfloat<6,2>, cfloat<4,1>
